src/streamlit_extras/prometheus/__init__.py

```diff
@@ -0,0 +1,136 @@
+from typing import List, NamedTuple
+
+from prometheus_client import CollectorRegistry
+from prometheus_client.openmetrics.exposition import generate_latest
+from streamlit.runtime.stats import CacheStatsProvider
+
+from .. import extra
+
+
+class CustomStat(NamedTuple):
+    metric_str: str = ""
+
+    def to_metric_str(self) -> str:
+        return self.metric_str
+
+    def marshall_metric_proto(self, metric) -> None:
+        """Custom OpenMetrics collected via protobuf is not currently supported."""
+        # Included to be compatible with the RequestHandler's _stats_to_proto() method:
+        # https://github.com/streamlit/streamlit/blob/1.29.0/lib/streamlit/web/server/stats_request_handler.py#L73
+        # Fill in dummy values so protobuf format isn't broken
+        label = metric.labels.add()
+        label.name = "cache_type"
+        label.value = "custom_metrics"
+
+        label = metric.labels.add()
+        label.name = "cache"
+        label.value = "not_implemented"
+
+        metric_point = metric.metric_points.add()
+        metric_point.gauge_value.int_value = 0
+
+
+class PrometheusMetricsProvider(CacheStatsProvider):
+    def __init__(self, registry: CollectorRegistry):
+        self.registry = registry
+
+    def get_stats(self) -> List[CustomStat]:
```

Comment:
In streamlit we expected `get_stats` to return `List[CacheStats]`, and to be compatible with a mechanism in streamlit stats manager get_stats, individual providers should return CacheStats that contain "category_name" and "cache_name" fields

Reply:
OK, as we discussed seems like that change will break this totally. So we'll want to wait until a fix in 1.31 until we can release this.